### DISCUSSION TITLE
Need help debugging this error: [FAIL. Reason: Setup failed: Index out of bounds]. Lesson 12.

### DISCUSSION THREAD

**Discussion by alfheimrShiven** [Cyfrin/foundry-full-course-cu#789] [Q&A] [answered]
In the defi-stablecoin project, the `DSCEngineTest.t.sol` setUp() function seems to be throwing the following error:

```zsh
shiven@Shivendras-MacBook-Pro defi-stablecoin % forge test -vvvv
[⠆] Compiling...
No files changed, compilation skipped
2023-09-04T08:59:40.802285Z ERROR forge::runner: setUp failed reason="Index out of bounds" contract=0x7fa9385be102ac3eac297483dd6233d62b3e1496

Running 1 test for test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: Setup failed: Index out of bounds] setUp() (gas: 0)
Traces:
  [8578093] DSCEngineTest::setUp() 
    ├─ [4074260] → new DeployDSC@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← 20127 bytes of code
    ├─ [4466780] DeployDSC::run() 
    │   ├─ [3431595] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast() 
    │   │   │   └─ ← ()
    │   │   ├─ [372255] → new MockV3Aggregator@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← 1082 bytes of code
    │   │   ├─ [372255] → new MockV3Aggregator@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   └─ ← 1082 bytes of code
    │   │   ├─ [590124] → new ERC20Mock@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← 2722 bytes of code
    │   │   ├─ [590124] → new ERC20Mock@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   └─ ← 2722 bytes of code
    │   │   ├─ [0] VM::stopBroadcast() 
    │   │   │   └─ ← ()
    │   │   └─ ← 6088 bytes of code
    │   ├─ [912] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← MockV3Aggregator: [0x90193C961A926261B756D1E5bb255e67ff9498A1], MockV3Aggregator: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496], ERC20Mock: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], ERC20Mock: [0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76], 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>) 
    │   │   └─ ← ()
    │   ├─ [749823] → new DecentralizedStableCoin@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← 3401 bytes of code
    │   ├─ [48491] → new <Unknown>@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← 0 bytes of code
    │   └─ ← "Index out of bounds"
    └─ ← "Index out of bounds"

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 5.82ms
 
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/DSCEngineTest.t.sol:DSCEngineTest
[FAIL. Reason: Setup failed: Index out of bounds] setUp() (gas: 0)

Encountered a total of 1 failing tests, 0 tests succeeded
shiven@Shivendras-MacBook-Pro defi-stablecoin % 
```

[My project repo](https://github.com/alfheimrShiven/defi-stablecoin)

Let me know if you spot the bug. Thanks in advance.

**Comment by theirrationalone** [comment]
Maybe you haven't updated your repo with your latest code. But your log traces are telling much... 

You can see, `DecentralizedStableCoin` is transferring ownership before `DSCEnigine` creation and deployment... And it's transferring ownership to an unknown address which is not an address of `DSCEngine`.

After that, your traces are telling that an unknown contract was being deployed after transferring ownership. 

There's no proof in your traces that you have deployed `DSCEngine` before transferring ownership of `DecentralizedStableCoin` to `DSCEngine`.

Make sure your repo is up to date and you're deploying a valid `DSCEngine` beforr transferring ownership of `DecentralizedStableCoin`.


```log
│ ├─ [749823] → new DecentralizedStableCoin@0x5FbDB2315678afecb367f032d93F642f64180aa3 │ │ ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266) │ │ └─ ← 3401 bytes of code │ ├─ [48491] → new <Unknown>@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512 │ │ └─ ← 0 bytes of code │ └─ ← "Index out of bounds" └─ ← "Index out of bounds"
```

**Reply by alfheimrShiven** [reply]
Hey @theirrationalone it would be helpful if you could point out the error in my repo. Your answer is not very specific. My DeployDSC.s.sol is exactly the same as the source code.

**Reply by theirrationalone** [reply]
@alfheimrShiven !, Somehow your DSCEngine deployment is getting `Unrecognized`. That's why its constructor throwing `Index out of bound`. Sorry buddy i don't have my machine yet that's why i'm unable to help you at this time with this issue. 

Hope @shikhar229169 !, @usmanfarooq91 !, @Neftyr !, may help you with this.... ! 😅

**Reply by shikhar229169** [reply, marked as answer]
Here is the error: https://github.com/alfheimrShiven/defi-stablecoin/blob/main/src/DSCEngine.sol#L117

The array is of zero length initially, so you can't use index for that.
Use push function to add collateral addresses to the arrray.

**Reply by Niferu** [reply]
Great catch @shikhar229169 !

**Reply by theirrationalone** [reply]
Woooww @shikhar229169 ! == Smart inspector 🕵

**Comment by theirrationalone** [comment]
> Great catch @shikhar229169 !
@Neftyr ! Thanks For the Response :`)

**Comment by alfheimrShiven** [comment]
> Here is the error: https://github.com/alfheimrShiven/defi-stablecoin/blob/main/src/DSCEngine.sol#L117
> 
> The array is of zero length initially, so you can't use index for that. Use push function to add collateral addresses to the array.


@shikhar229169 Thanks bro for pointing it out. In my head, I was considering s_collateralTokens a mapping instead of it just being an array, hence ended up adding the token addresses to it using the assignment operator. Dang!
Appreciate your support guys! @theirrationalone @shikhar229169 @Neftyr 🫡
